external/swloop/swloop.cpp

```diff
@@ -0,0 +1,533 @@
+    MatrixXT<T> res = MatrixXT<T>::Zero(sz1, sz2);
+    for (int ii=0; ii<ind.rows(); ii++) {
+        // Indices are direct from Matlab so use base-1 and are also doubles not ints
+        res(static_cast<size_t>(ind(ii, 0)) - 1, static_cast<size_t>(ind(ii, 1)) - 1) += data(ii);
+    }
+    return res;
+}
+
+std::atomic_int err_flag(0);
+std::vector<std::string> errmsgs = {
+    std::string("all good"),
+    std::string("swloop:notposdef: Very baaaad!"),
+    std::string("swloop:notposdef: The input matrix is not positive definite."),
+    std::string("swloop: Eigensolver could not converge")
+};
+
+void swcalc_fun(size_t i0, size_t i1, struct pars &params, struct swinputs &inputs, struct swoutputs &outputs) {
+    // Setup
+    size_t nHam = 2 * params.nMagExt;
+    size_t len_abcd = 3 * params.nBond;
+    // Assuming all data is column-major(!), so can be mapped directly to Eigen matrices
+    // We're also assuming that the memory blocks are actually as big as defined here!
+    // (E.g. the code which calls this needs to do some bounds checking)
+    Eigen::Map<const Eigen::MatrixXd> hklExt(inputs.hklExt, 3, params.nHkl);
+    Eigen::Map<const Eigen::VectorXcd> ABCD (inputs.ABCD, len_abcd);
+    Eigen::Map<const Eigen::MatrixXd> idxAll (inputs.idxAll, len_abcd, 2);
+    Eigen::Map<const Eigen::VectorXd> ham_diag (inputs.ham_diag, nHam);
+    Eigen::Map<const RowMatrixXd> dR (inputs.dR, params.nBond, 3);
+    Eigen::Map<const Eigen::MatrixXd> RR (inputs.RR, 3, params.nMagExt);
+    Eigen::VectorXd sqrtS0 = Eigen::VectorXd(params.nMagExt);
+    for (int ii=0; ii<sqrtS0.rows(); ii++) {
+        sqrtS0(ii) =  sqrt(inputs.S0[ii] / 2.); }
+    Eigen::Map<const Eigen::MatrixXcd> zed (inputs.zed, 3, params.nMagExt);
+    Eigen::Map<const Eigen::MatrixXd> FF (inputs.FF, params.nMagExt, params.nHkl);
+    Eigen::Map<const RowMatrixXd> bqdR (inputs.bqdR, params.nBqBond, 3);
+    Eigen::Map<const Eigen::VectorXd> bqABCD (inputs.bqABCD, 3 * params.nBqBond);
+    Eigen::Map<const Eigen::MatrixXd> idxBq (inputs.idxBq, 3 * params.nBqBond, 2);
+    Eigen::Map<const Eigen::VectorXd> bq_ham_diag (inputs.bq_ham_diag, nHam);
+    std::vector< Eigen::Map<const Eigen::MatrixXd> > ham_MF;
+    if (params.field) {
+        for (size_t ii=0; ii<params.nTwin; ii++) {
+            ham_MF.push_back(Eigen::Map<const Eigen::MatrixXd>(inputs.ham_MF_v[ii], nHam, nHam));
+        }
+    }
+
+    Eigen::MatrixXd gComm = Eigen::MatrixXd::Identity(nHam, nHam);
+    for (size_t ii=params.nMagExt; ii<nHam; ii++) {
+        gComm(ii, ii) = -1.; }
+    Eigen::Map<const Eigen::VectorXi> idx0(inputs.idx0, params.nHkl);
+    size_t nHklT = params.nHkl / params.nTwin;
+
+    std::complex<double> *omega = outputs.omega;
+    std::complex<double> *Sab_ptr = outputs.Sab;
+
+    size_t nHklI = params.nHkl / params.nTwin / 3;
+    Eigen::Matrix3cd K1, K2, cK1, nx, m1, K;
+    Eigen::Vector3d n;
+    if (params.incomm) {
+        // Defines the Rodrigues rotation matrix to rotate Sab back
```

Comment:
Arguably would be keep the main block more concise to have this in a separate function even though it's only used once - happy for you to push back on this as very nit-picking!